lib/bitcoin/MongoDbBlockMetadataStore.ts

```diff
@@ -0,0 +1,116 @@
+   */
+  constructor (serverUrl: string, databaseName?: string) {
+    super(serverUrl, MongoDbBlockMetadataStore.collectionName, databaseName);
+  }
+
+  protected async createIndex (collection: Collection) {
+    // Create unique index, so duplicate inserts are rejected.
+    await collection.createIndex({ height: 1 }, { unique: true });
+  }
+
+  public async add (arrayOfBlockMetadata: BlockMetadata[]): Promise<void> {
+    const bulkOperations = this.collection!.initializeOrderedBulkOp();
+
+    arrayOfBlockMetadata.sort((a, b) => a.height - b.height);
+    for (const blockMetadata of arrayOfBlockMetadata) {
+      bulkOperations.find({ height: blockMetadata.height }).upsert().replaceOne(blockMetadata);
+    }
+
+    await bulkOperations.execute();
+  }
+
+  public async removeLaterThan (blockHeight?: number) {
+    // If block height is not given, remove all.
+    if (blockHeight === undefined) {
+      await this.clearCollection();
+      return;
+    }
+
+    await this.collection!.deleteMany({ height: { $gt: blockHeight } });
+  }
+
+  public async get (fromInclusiveHeight: number, toExclusiveHeight: number): Promise<BlockMetadata[]> {
+    let dbCursor: Cursor<BlockMetadata>;
+
+    // Add filter to query.
+    dbCursor = this.collection!.find({ $and: [
+      { height: { $gte: fromInclusiveHeight } },
+      { height: { $lt: toExclusiveHeight } }
+    ] });
+
+    // Add sort to query.
+    dbCursor = dbCursor.sort({ height: 1 });
+
+    // Execute the query.
+    const blocks = await dbCursor.toArray();
+    return blocks;
+  }
+
+  public async getLast (): Promise<BlockMetadata | undefined> {
+    const blocks = await this.collection!.find<BlockMetadata>().sort({ height: -1 }).limit(1).toArray();
+    if (blocks.length === 0) {
+      return undefined;
+    }
+
+    const lastBlockMetadata = blocks[0];
+    return lastBlockMetadata;
+  }
+
+  /**
+   * Gets the first block (block with lowest height).
+   */
+  private async getFirst (): Promise<BlockMetadata | undefined> {
+    const blocks = await this.collection!.find<BlockMetadata>().sort({ height: 1 }).limit(1).toArray();
+    if (blocks.length === 0) {
+      return undefined;
+    }
+
+    const lastBlockMetadata = blocks[0];
+    return lastBlockMetadata;
+  }
+
+  async lookBackExponentially (): Promise<BlockMetadata[]> {
+    const lastBlock = await this.getLast();
+    const firstBlock = await this.getFirst();
+
+    if (firstBlock === undefined) {
+      return [];
+    }
+
+    // Exponentially look back from last block to first block.
+    const heightOfBlocksToReturn: number[] = [];
+    let lookBackDistance = 1;
+    let currentHeight = lastBlock!.height;
+    while (currentHeight >= firstBlock.height) {
+      heightOfBlocksToReturn.push(currentHeight);
+
+      currentHeight = lastBlock!.height - lookBackDistance;
+      lookBackDistance *= 2;
+    }
+
+    const exponentiallySpacedBlocks = await this.collection!.find<BlockMetadata>(
```

Comment:
Lint warning that this `find` function w/ the 2nd option is deprecated?

Reply:
Had to update the mongodb types version :)

Reply:
You need to run `npm i`. This is the reason why MongoDB type version was updated.